src/codegen/mod.rs

```diff
@@ -1584,97 +1584,44 @@ impl CodeGenerator for Function {
     }
 }
 
+// Return true if any of the ancestors of `id` are in the whitelisted items set,
+// false otherwise.
+fn ancestor_is_whitelisted(ctx: &BindgenContext, whitelisted_items: &ItemSet, id: ItemId) -> bool {
+    let item = ctx.resolve_item(id);
+    let mut last = id;
+    let mut current = item.parent_id();
+
+    while last != current {
+        if whitelisted_items.contains(&current) {
+            return true;
+        }
+        last = current;
+        current = ctx.resolve_item(current).parent_id();
+    }
+
+    false
+}
+
 pub fn codegen(context: &mut BindgenContext) -> Vec<P<ast::Item>> {
     context.gen(|context| {
         let mut result = CodegenResult::new();
 
         debug!("codegen: {:?}", context.options());
 
-        // If the whitelisted types and functions sets are empty, just generate
-        // everything.
-        if context.options().whitelisted_types.is_empty() &&
-           context.options().whitelisted_functions.is_empty() &&
-           context.options().whitelisted_vars.is_empty() {
-            for (_item_id, item) in context.items() {
-                // Non-toplevel item parents are the responsible one for generating
-                // them.
-                if item.is_toplevel(context) {
-                    item.codegen(context, &mut result, &());
-                }
-            }
-        } else {
-            // Recursively collect all the types dependent on the whitelisted
-            // types, then generate them.
-            //
-            // FIXME(emilio): This pass is probably slow, but it can't be faster
-            // than docopt anyway :)
-            let mut items = ItemSet::new();
-            for (_item_id, item) in context.items() {
-                // FIXME(emilio): This probably should look only at whether the
-                // parent is a module.
-                if !item.is_toplevel(context) {
-                    continue;
-                }
+        let whitelisted_items: ItemSet = context.whitelisted_items().collect();
```

Comment:
Yes, because checking if the ancestor is in the set requires that we have fully computed the set.